hutch_python/ipython_session_timer.py

```diff
@@ -0,0 +1,93 @@
+"""
+This module modifies an ``ipython`` shell to automatically close if it has been 
+idle for a certain number of hours. If no command is entered in the ipython instance
+for more than the maximum idle time, the session is automatically closed. The 
+maximum idle time can be updated. Currently, it is set to 48 hours.
+"""
+
+import time
+from IPython import get_ipython
+from threading import Thread
+
+
+class IPythonSessionTimer:
+    '''
+    Class tracks the amount of time the current `InteractiveShell` instance (henceforth
+    called 'user session') has been idle and closes the session if more than 48
+    hours have passed.
+
+    Time is in seconds (floating point) since the epoch began. (In UNIX the
+    epoch started on January 1, 1970, 00:00:00 UTC)
+
+    Parameters
+    ----------
+    ipython : ``IPython.terminal.interactiveshell.TerminalInteractiveShell``
+        The active ``ipython`` ``Shell``, perhaps the one returned by
+        ``IPython.get_ipython()``.
+
+    Attributes
+    ----------
+    curr_time: float
+        The current time in seconds.
+
+    max_idle_time: int
+        The maximum number of seconds a user session can be idle (currently set 
+        to 172800 seconds or 48 hours).
+    
+    last_active_time: float
+        The time of the last user activity in this session.
+
+    idle_time: float
+        The amount of time the user session has been idle.
+    '''
+
+    def __init__(self, ipython):
+        self.curr_time = 0
+        self.max_idle_time = 172800
+        self.last_active_time = 0
+        self.idle_time = 0
+
+        # _set_last_active_time() function will trigger every time user runs a cell
+        ipython.events.register('post_run_cell', self._set_last_active_time)
+
+    def _set_last_active_time(self, result):
+        self.last_active_time = time.time()
+
+    def _get_time_passed(self):
+        self.curr_time = time.time()
+        self.idle_time = self.curr_time - self.last_active_time
+
+    def _timer(self, sleep_time):
+        time.sleep(sleep_time)
+
+    def _start_session(self):
+
+        # Check if idle_time has exceeded max_idle_time
+        while (self.idle_time < self.max_idle_time):
+            self._timer(self.max_idle_time - self.idle_time)
+            self._get_time_passed()
+
+        # Close the user session
+        print("This hutch-python session has timed out. Please start a new session.")
+
+        # Close this ipython session
+        get_ipython().ask_exit()
```

Comment:
I'm also not sure if there's a better way.  Currently if the session times out, we can actually still request a close with `Ctrl+D`, though it doesn't matter what we choose.
![image](https://github.com/pcdshub/hutch-python/assets/35379409/35ac151c-5433-47e7-9214-eaceca3d1669)

Reply:
Taking inspiration from the linked github thread, we apparently need to follow this up with another exit call to kill the prompt that's waiting for input, otherwise as you say in the desc this waits for user input before exiting.
```
ip = get_ipython()
ip.ask_exit()
ip.pt_app.app.exit()
```